train_gpt2.cu

```diff
@@ -919,14 +963,50 @@ void gpt2_backward_and_reduce(GPT2 *model, int* inputs, const int* targets, int
                 4 * C * C, 4 * C,
                 C * 4 * C, C
             };
+            if(multi_gpu_config.zero_stage == 2) {
+#if MULTI_GPU
+                // wait for last layer's transfer to be completed, so we don't get any race conditions
+                // where main stream is writing to a buffer that nccl stream is reading
+                // this will result in a structure as follows:
+                // Main stream:     Calculate grads of layer 1 in buffer 1
+                // NCCL stream:     Wait for main stream
+                // Main stream:     Calculate grads of layer 2 in buffer 2
+                // NCCL stream:     Transmit data from buffer 1
+                // SYNC NNCL stream
+                // NCCL stream:     Wait for buffer 2 to be ready
+                // Main stream:     calculate grads of layer 3 in buffer 1
+                // ...
+                cudaCheck(cudaStreamSynchronize(multi_gpu_config.nccl_stream));
```

Comment:
cudaStreamSynchronize will wait on the host which isn't great for performance/overlap, would it be possible to synchronise with events between streams, similar to what we do else with:

    cudaCheck(cudaEventRecord(config->compute_nccl_sync, compute_stream));
    cudaCheck(cudaStreamWaitEvent(config->nccl_stream, config->compute_nccl_sync));